clarity/src/vm/ast/parser_v2/mod.rs

```diff
@@ -0,0 +1,3227 @@
+            message: e.e.message(),
+            spans: vec![e.span.clone()],
+            suggestion: None,
+        })
+        .collect();
+    (stmts, diagnostics, parser.success)
+}
+
+pub fn decode_hex(s: &str) -> Result<Vec<u8>, ParseIntError> {
+    (0..s.len())
+        .step_by(2)
+        .map(|i| u8::from_str_radix(&s[i..i + 2], 16))
+        .collect()
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::vm::{
+        diagnostic::Level,
+        types::{
+            ASCIIData, CharType, PrincipalData, SequenceData, StandardPrincipalData, UTF8Data,
+        },
+    };
+
+    use super::*;
+
+    #[test]
+    fn test_parse_int() {
+        let (stmts, diagnostics, success) = parse_collect_diagnostics("    123");
+        assert_eq!(success, true);
+        assert_eq!(stmts.len(), 1);
+        assert_eq!(diagnostics.len(), 0);
+        if let Some(Value::Int(123)) = stmts[0].match_atom_value() {
+        } else {
+            panic!("failed to parse int value");
+        }
+        assert_eq!(
+            stmts[0].span,
+            Span {
+                start_line: 1,
+                start_column: 5,
+                end_line: 1,
+                end_column: 7
+            }
+        );
+
+        let (stmts, diagnostics, success) = parse_collect_diagnostics("    -123");
+        assert_eq!(success, true);
+        assert_eq!(stmts.len(), 1);
+        assert_eq!(diagnostics.len(), 0);
+        if let Some(Value::Int(-123)) = stmts[0].match_atom_value() {
+        } else {
+            panic!("failed to parse negative int value");
+        }
+        assert_eq!(
+            stmts[0].span,
+            Span {
+                start_line: 1,
+                start_column: 5,
+                end_line: 1,
+                end_column: 8
+            }
+        );
+
+        let (stmts, diagnostics, success) = parse_collect_diagnostics("42g ");
+        assert_eq!(success, false);
+        assert_eq!(stmts.len(), 1);
+        if let Some("42g") = stmts[0].match_placeholder() {
+        } else {
+            panic!("failed to parse int value with error");
+        }
+        assert_eq!(
+            stmts[0].span,
+            Span {
+                start_line: 1,
+                start_column: 1,
+                end_line: 1,
+                end_column: 3
+            }
+        );
+        assert_eq!(diagnostics.len(), 1);
+        assert_eq!(diagnostics[0].level, Level::Error);
+        assert_eq!(
+            diagnostics[0].message,
+            "invalid character, 'g', in int literal".to_string()
+        );
+        assert_eq!(
+            diagnostics[0].spans[0],
+            Span {
+                start_line: 1,
+                start_column: 3,
+                end_line: 1,
+                end_column: 3
+            }
+        );
+
+        let (stmts, diagnostics, success) =
+            parse_collect_diagnostics("340282366920938463463374607431768211456 ");
+        assert_eq!(success, false);
```

Comment:
This fails due to numeric overflow, right?

Reply:
Yes, this number exceeds the range of a 128-bit integer. I'll add a comment.

Reply:
5525fcead